doc/contrib/unit_tests.rst

```diff
@@ -0,0 +1,232 @@
+To write a test for Scala, see `Scalatest tutorial <http://www.scalatest.org/user_guide/writing_your_first_test>`_.
+
+You may try running your test by following instructions in :ref:`this section <running_jvm_tests>`.
+
+R package: testthat
+===================
+Add your test under the directory `R-package/tests/testthat <https://github.com/dmlc/xgboost/tree/master/R-package/tests/testthat>`_. Refer to `this excellent tutorial on testthat <https://kbroman.org/pkg_primer/pages/tests.html>`_.
+
+You may try running your test by following instructions in :ref:`this section <running_r_tests>`.
+
+**************************
+Running Unit Tests Locally
+**************************
+
+.. _running_r_tests:
+
+R package
+=========
+Run
+
+.. code-block:: bash
+
+  make Rcheck
+
+at the root of the project directory.
+
+.. _running_jvm_tests:
+
+JVM packages
+============
+As part of the building process, tests are run:
+
+.. code-block:: bash
+
+  mvn package
+
+.. _running_pytest:
+
+Python package: pytest
+======================
+
+.. note:: Having issue? Try Docker container
+
+  If you are having difficulty running the commands below (e.g. due to missing packages), consider using our Docker container. See :ref:`running_tests_inside_docker`.
+
+To run Python unit tests, first install `pytest <https://docs.pytest.org/en/latest/contents.html>`_ package:
+
+.. code:: bash
+
+  pip3 install --user pytest
+
+Then compile XGBoost:
+
+.. code:: bash
+
+  mkdir build
+  cd build
+  cmake ..
+  make
+  cd ..
+
+Now invoke pytest at the project root directory:
+
+.. code:: bash
+
+  export PYTHONPATH=./python-package
+  pytest -v -s --fulltrace tests/python
+
+In addition, to build and test CUDA code, run:
+
+.. code:: bash
+
+  cd build
+  cmake -DUSE_CUDA=ON -DUSE_NCCL=ON ..
+  make
+  cd ..
+
+  pytest -v -s --fulltrace tests/python-gpu
+
+.. _running_gtest:
+
+C++: Google Test
+================
+
+.. note:: Having issue? Try Docker container
+
+  If you are having difficulty running the commands below (e.g. due to missing packages), consider using our Docker container. See :ref:`running_tests_inside_docker`.
+
+To build and run C++ unit tests, install `Google Test <https://github.com/google/googletest>`_ library with headers
+and then enable tests while running CMake:
+
+.. code-block:: bash
+
+  mkdir build
+  cd build
+  cmake -DGOOGLE_TEST=ON -DGTEST_ROOT=/path/to/google-test ..
+  make
+  make test
+
+To enable tests for CUDA code, add ``-DUSE_CUDA=ON`` and ``-DUSE_NCCL=ON`` (CUDA toolkit required):
+
+.. code-block:: bash
+
+  mkdir build
+  cd build
+  cmake -DGOOGLE_TEST=ON -DGTEST_ROOT=/path/to/google-test -DUSE_CUDA=ON -DUSE_NCCL=ON ..
+  make
+  make test
+
+One can also run all unit test using ctest tool which provides higher flexibility. For example:
+
+.. code-block:: bash
+
+  ctest --verbose
+
+.. _running_tests_inside_docker:
+
+Running tests inside a Docker container (Recommended)
```

Comment:
Actually having people to fail on tests are good thing.  Many occasions some bugs that weren't caught on test farm showed up on my machine, or sometimes the other way around.  Encouraging diversity of environment may help us catching more bugs.  I don't see the necessity of promoting Docker since we have a fairly simple dependency tree.

Reply:
Okay, let me go ahead and remove the Docker instructions.